der/src/document.rs

```diff
@@ -71,49 +106,99 @@ pub trait Document<'a>: AsRef<[u8]> + Sized + TryFrom<Vec<u8>, Error = Error> {
     /// Encode ASN.1 DER document as a PEM string.
     #[cfg(feature = "pem")]
     #[cfg_attr(docsrs, doc(cfg(feature = "pem")))]
-    fn to_pem(&self, line_ending: pem::LineEnding) -> Result<String>
+    pub fn to_pem(&self, line_ending: pem::LineEnding) -> Result<String>
     where
-        Self: pem::PemLabel,
+        T: PemLabel,
     {
         Ok(pem::encode_string(
-            Self::TYPE_LABEL,
+            T::TYPE_LABEL,
             line_ending,
-            self.as_ref(),
+            self.as_der(),
         )?)
     }
 
     /// Read ASN.1 DER document from a file.
     #[cfg(feature = "std")]
     #[cfg_attr(docsrs, doc(cfg(feature = "std")))]
-    fn read_der_file(path: impl AsRef<Path>) -> Result<Self> {
+    pub fn read_der_file(path: impl AsRef<Path>) -> Result<Self>
+    where
+        T: for<'a> Decodable<'a> + Sized,
+    {
         fs::read(path)?.try_into()
     }
 
+    /// Write ASN.1 DER document to a file.
+    #[cfg(feature = "std")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "std")))]
+    pub fn write_der_file(&self, path: impl AsRef<Path>) -> Result<()> {
+        write_file(path, self.as_der(), T::SENSITIVE)
+    }
+
     /// Read PEM-encoded ASN.1 DER document from a file.
     #[cfg(all(feature = "pem", feature = "std"))]
     #[cfg_attr(docsrs, doc(cfg(all(feature = "pem", feature = "std"))))]
-    fn read_pem_file(path: impl AsRef<Path>) -> Result<Self>
+    pub fn read_pem_file(path: impl AsRef<Path>) -> Result<Self>
     where
-        Self: pem::PemLabel,
+        T: for<'a> Decodable<'a> + PemLabel + Sized,
     {
         Self::from_pem(&fs::read_to_string(path)?)
     }
 
-    /// Write ASN.1 DER document to a file.
-    #[cfg(feature = "std")]
-    #[cfg_attr(docsrs, doc(cfg(feature = "std")))]
-    fn write_der_file(&self, path: impl AsRef<Path>) -> Result<()> {
-        write_file(path, self.as_ref(), Self::SENSITIVE)
-    }
-
     /// Write PEM-encoded ASN.1 DER document to a file.
     #[cfg(all(feature = "pem", feature = "std"))]
     #[cfg_attr(docsrs, doc(cfg(all(feature = "pem", feature = "std"))))]
-    fn write_pem_file(&self, path: impl AsRef<Path>, line_ending: pem::LineEnding) -> Result<()>
+    pub fn write_pem_file(&self, path: impl AsRef<Path>, line_ending: pem::LineEnding) -> Result<()>
     where
-        Self: pem::PemLabel,
+        T: PemLabel,
     {
-        write_file(path, self.to_pem(line_ending)?.as_bytes(), Self::SENSITIVE)
+        write_file(path, self.to_pem(line_ending)?.as_bytes(), T::SENSITIVE)
+    }
+
+    /// Attempt to decode `self.der_bytes` as `T`.
+    ///
+    /// This method doesn't uphold the invariant that `T` always decodes
+    /// successfully, but is needed to make the lifetimes for the constructor
+    /// work.
+    fn try_decode<'a>(&'a self) -> Result<T>
+    where
+        T: Decodable<'a> + Sized,
+    {
+        T::from_der(self.as_der())
+    }
```

Comment:
I think we should ditch that invariant altogether.

Reply:
`Document` combines three functionalities:
1. Zeroing the `Vec<u8>`. This has clear value.
2. Ensuring file permissions on write. This has clear value. I have some minor concern, however, that this is only useful in a narrow set of cases (i.e. where you want to do full, synchronous io). If you want to do partial io (such as a certificate chain incremental io) or asynchronous io, this type drops this value.
3. Ensuring the bytes will successfully decode to `T`. The value of this is unclear to me. And it carries a performance and security cost.

Dropping the third guarantee still leaves `Document` with two features. So it is still better than a `Vec<u8>`.